lab-maddy/server.js

```diff
@@ -0,0 +1,62 @@
+'use strict';
+
+const http = require('http');
+const url = require('url');
+const queryString = require('querystring');
+
+const cowsay = require('cowsay');
+const PORT = process.env.PORT || 3000;
+
+const bodyParse = function(req, callback){
+  if(req.method === 'POST' || req.method === 'PUT') {
+    let body = '';
+    req.on('data', buffer => {
+      body += buffer.toString();
+    });
+    req.on('end', () => callback(null, body));
+    req.on('error', callback);
+  } else {
+    callback(null, '{}');
+  }
+};
+
+const server = module.exports = http.createServer((req, res)=> {
+  req.url = url.parse(req.url);
+  req.url.query = queryString.parse(req.url.query);
+  bodyParse(req, (err, body)=> {
+    if(err){
+      res.writeHead(500);
+      res.end();
+      return;
+    }
+
+    try {
+      req.body = JSON.parse(body);
+    } catch(e) {
+      res.writeHead(400);
+      res.end();
+      return;
+    }
+
+    if(req.method === 'GET' && req.url.pathname === '/cowsay') {
+      // console.log(JSON.stringify(req.url.query));
+      if(!req.url.query.text) {
+        res.writeHead(400, {
+          'Content-Type': 'text/plain',
+        });
+        res.write(cowsay.say({ text: 'bad request'}));
+        res.end();
+        return;
+      } else {
+        res.writeHead(200, {
+          'Content-Type': 'text/plain',
+        });
+        res.write(cowsay.say(req.url.query));
+        res.end();
+        return;
+      }
+    }
+  });
+});
+
+server.listen(PORT, ()=> console.log(`listening on: ${PORT}`));
```

Comment:
Looks like you need to still add routes for the '/' path. This will behave the same regardless of what method they use. Additionally, you will also need to add a route for POST on '/cowsay'. Because you don't have this yet is probably why your first test is failing. It will be very similar to the GET route except you wont be dealing with the url query, instead you will be dealing with the req.body.